ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -176,6 +160,29 @@ where
     }
 }
 
+/// Computes the dot product of two arrays of the same size.
+/// It is isolated from Lagrange because there could be potential SIMD optimizations used
+fn dot_product<F: PrimeField, const N: usize>(a: &[F; N], b: &[F; N]) -> F {
+    // Staying in integers allows rustc to optimize this code properly
+    // with any reasonable N, we won't run into overflow with dot product.
+    // (N can be as large as 2^32 and still no chance of overflow for 61 bit prime fields)
+    debug_assert!(
+        F::PRIME.into() < (1 << 64),
+        "The prime {} is too large for this dot product implementation",
+        F::PRIME.into()
+    );
+
+    let mut sum = 0;
+
+    // I am cautious about using zip in hot code
+    // https://github.com/rust-lang/rust/issues/103555
+    for i in 0..N {
+        sum += a[i].as_u128() * b[i].as_u128();
+    }
+
+    F::truncate_from(sum)
```

Comment:
How does this work? It looks like truncate_from() uses the argument as a seed to map to a pseudorandom field element. Are there correctness tests for this?

Reply:
that would be very bad - this code just defers modulo reduction operation till the very end: https://github.com/private-attribution/ipa/blob/bef33119b463c3cbd34f82a962f00cc3063055d0/ipa-core/src/ff/prime_field.rs#L130

Reply:
hmm, can you please point me to `Fp25519::truncate_from`? I couldn't find it and it would be also bad if we had `truncate_from` implementation for it like you described.

>is it possible to use try_from instead

`TryFrom` wouldn't work here because it will reject values > PRIME: https://github.com/private-attribution/ipa/blob/bef33119b463c3cbd34f82a962f00cc3063055d0/ipa-core/src/ff/prime_field.rs#L226

Reply:
If this line is always doing a modulo reduction to get back to the field then that's fine. Truncate to me implies dropping bits, but is this just how the modulo-reduce-into-field function is named?

Reply:
The function I was looking at was this one: https://github.com/akoshelev/raw-ipa/blob/8bb84e5ae9acfd529e482d8663c44ea96e64ea27/src/ff/ec_prime_field.rs#L176

Reply:
I see - this looks like old code, we don't have it in main: https://github.com/private-attribution/ipa/blob/main/ipa-core/src/ff/ec_prime_field.rs

Reply:
> Truncate to me implies dropping bits, but is this just how the modulo-reduce-into-field function is named?
Agree, this got very confusing over time. Originally it was dropping bits but we changed it many times and here we are.

Let me open an issue on this one, I agree that naming is important

Reply:
https://github.com/private-attribution/ipa/issues/1421